File: carrera-sdk/producer/java/carrera-producer-sdk/src/main/java/com/xiaojukeji/carrera/nodemgr/connection/CarreraConnection.java
package com.xiaojukeji.carrera.nodemgr.connection;

import com.xiaojukeji.carrera.nodemgr.Node;
import com.xiaojukeji.carrera.producer.CarreraReturnCode;
import com.xiaojukeji.carrera.thrift.*;
import org.apache.thrift.protocol.TCompactProtocol;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.protocol.TProtocolException;
import org.apache.thrift.transport.TFramedTransport;
import org.apache.thrift.transport.TSSLTransportFactory;
import org.apache.thrift.transport.TSocket;
import org.apache.thrift.transport.TTransport;
import org.apache.thrift.transport.TTransportException;

import java.util.List;


public class CarreraConnection {

    private ProducerService.Client thriftClient;
    private Node node;
    private int clientTimeout;
    private TSocket socket;

    private static final boolean OPEN_SSL = Boolean.valueOf(System.getProperty("carrera.ssl.open", "false"));
    private static final String TRUST_STORE = System.getProperty("carrera.ssl.trustStore", "");
    private static final String TRUST_PASS = System.getProperty("carrera.ssl.trustPass", "");


    public CarreraConnection(Node node, int clientTimeout) {
        this.node = node;
        this.clientTimeout = clientTimeout;
    }

    private ProducerService.Client createThriftClient() throws TTransportException {
        TSocket socket;
        if (OPEN_SSL) {
            TSSLTransportFactory.TSSLTransportParameters parameters = new TSSLTransportFactory.TSSLTransportParameters();
            parameters.setTrustStore(TRUST_STORE, TRUST_PASS);
            socket = TSSLTransportFactory.getClientSocket(node.getIp(), node.getPort(), clientTimeout, parameters);
        } else {
            socket = new TSocket(node.getIp(), node.getPort(), clientTimeout);
            socket.open();
        }

        TTransport transport = new TFramedTransport(socket);
        TProtocol protocol = new TCompactProtocol(transport);
        return new ProducerService.Client(protocol);
    }

    public synchronized Result send(Message msg, long proxyTimeout) {
        Result ret = new Result();
        try {
            if (this.thriftClient == null) {
                this.thriftClient = createThriftClient();
            }
            ret = thriftClient.sendSync(msg, proxyTimeout);

        } catch (TTransportException | TProtocolException te) {
            ret.setCode(CarreraReturnCode.THRIFT_NETWORK_EXCEPTION);
            ret.setMsg(te.getClass().getSimpleName() + " - " + te.getMessage());
            close();
        } catch (Exception te) {
            ret.setCode(CarreraReturnCode.THRIFT_EXCEPTION);
            ret.setMsg(te.getClass().getSimpleName() + " - " + te.getMessage());
            close();
        }
        return ret;
    }

    public synchronized DelayResult sendDelay(DelayMessage delayMessage, long proxyTimeout) {
        DelayResult result = new DelayResult();
        try {
            if (this.thriftClient == null) {
                this.thriftClient = createThriftClient();
            }
            result = thriftClient.sendDelaySync(delayMessage, proxyTimeout);

        } catch (TTransportException | TProtocolException te) {
            result.setCode(CarreraReturnCode.THRIFT_NETWORK_EXCEPTION);
            result.setMsg(te.getClass().getSimpleName() + " - " + te.getMessage());
            close();
        } catch (Exception te) {
            result.setCode(CarreraReturnCode.THRIFT_EXCEPTION);
            result.setMsg(te.getClass().getSimpleName() + " - " + te.getMessage());
            close();
        }
        return result;
    }

    public synchronized Result sendBatchSync(List<Message> messages) {
        Result ret = new Result();
        try {
            if (this.thriftClient == null) {
                this.thriftClient = createThriftClient();
            }
            ret = thriftClient.sendBatchSync(messages);

        } catch (TTransportException | TProtocolException te) {
            ret.setCode(CarreraReturnCode.THRIFT_NETWORK_EXCEPTION);
            ret.setMsg(te.getClass().getSimpleName() + " - " + te.getMessage());
            close();
        } catch (Exception te) {
            ret.setCode(CarreraReturnCode.THRIFT_EXCEPTION);
            ret.setMsg(te.getClass().getSimpleName() + " - " + te.getMessage());
            close();
        }
        return ret;
    }

    public synchronized void close() {
        if (socket != null) {
            socket.close();
            socket = null;
        }
        this.thriftClient = null;
    }

    public Node getNode() {
        return node;
    }

    public void setNode(Node node) {
        this.node = node;
    }

    public synchronized boolean validate() {
        return socket == null || !socket.isOpen() || this.thriftClient == null;
    }

}

File: carrera-sdk/producer/cpp/Producer/CarreraConnection.cpp
#include "CarreraConnection.h"
#include "CarreraDefine.h"

#include <vector>
#include <boost/shared_ptr.hpp>
#include <transport/TBufferTransports.h>
#include <transport/TSocket.h>
#include <protocol/TCompactProtocol.h>
#include <boost/algorithm/string.hpp>
#include <exception>
#include <iostream>

using namespace apache::thrift;
using namespace apache::thrift::protocol;
using namespace apache::thrift::transport;
using std::string;
using boost::shared_ptr;


namespace CarreraProducer {

    void CarreraConnection::Close() {
        apache::thrift::concurrency::Guard g(conn_mutex_);
        if (this->thrift_client_ != NULL) {
            (this->thrift_client_)->getInputProtocol()->getTransport()->close();
        }
        (this->thrift_client_).reset();
    }

    boost::shared_ptr<ProducerServiceClient> CarreraConnection::CreateThriftClient() {
        std::vector<std::string> ipport;
        boost::split(ipport, this->proxy_addr_, boost::is_any_of(":"));

        if (ipport.size() != 2) {
            return shared_ptr<ProducerServiceClient>();
        }
        std::string ip = ipport[0];
        int port = atoi(ipport[1].c_str());

        boost::shared_ptr<TSocket> socket(new TSocket(ip, port));
        boost::shared_ptr<TTransport> transport(new TFramedTransport(socket));

        socket->setConnTimeout(this->client_timeout_);
        socket->setRecvTimeout(this->client_timeout_);
        socket->setSendTimeout(this->client_timeout_);
        try {
            transport->open();
        } catch (TException &tx) {
            state_ = CONNECTION_UNHEALTHY;
            return boost::shared_ptr<ProducerServiceClient>();
        }

        boost::shared_ptr<TProtocol> protocol(new TCompactProtocol(transport));

        return boost::shared_ptr<ProducerServiceClient>(new ProducerServiceClient(protocol));
    }

    Result CarreraConnection::SendBatchSync(const std::vector<Message> &msgs) {
        Result ret;
        try {
            if (NULL == this->thrift_client_) {
                this->thrift_client_ = CreateThriftClient();
                if (NULL == this->thrift_client_) {
                    ret.__set_code(FAIL_SERVER_CONNECTION_ERROR);
                    ret.__set_msg("Illegal ipport or open transport failed.");
                    return ret;
                }
            }
            this->thrift_client_->sendBatchSync(ret, msgs);
        } catch (TException &e) {
            state_ = CONNECTION_UNHEALTHY;
            ret.__set_code(FAIL_SERVER_CONNECTION_ERROR);
            ret.__set_msg(e.what());
            (this->thrift_client_).reset();
        } catch (std::exception &e) {
            state_ = CONNECTION_UNHEALTHY;
            ret.__set_code(FAIL_SERVER_CONNECTION_ERROR);
            ret.__set_msg(e.what());
            (this->thrift_client_).reset();
        } catch (...) {
            state_ = CONNECTION_UNHEALTHY;
            ret.__set_code(FAIL_SERVER_CONNECTION_ERROR);
            ret.__set_msg("Unkown exception.");
            (this->thrift_client_).reset();
        }
        return ret;
    }

    Result CarreraConnection::Send(const Message &msg, int64_t proxy_timeout) {
        Result ret;
        try {
            if (NULL == this->thrift_client_) {
                this->thrift_client_ = CreateThriftClient();
                if (NULL == this->thrift_client_) {
                    ret.__set_code(FAIL_SERVER_CONNECTION_ERROR);
                    ret.__set_msg("Illegal ipport or open transport failed.");
                    return ret;
                }
            }
            this->thrift_client_->sendSync(ret, msg, proxy_timeout);
        } catch (TException &e) {
            state_ = CONNECTION_UNHEALTHY;
            ret.__set_code(FAIL_SERVER_CONNECTION_ERROR);
            ret.__set_msg(e.what());
            (this->thrift_client_).reset();
        } catch (std::exception &e) {
            state_ = CONNECTION_UNHEALTHY;
            ret.__set_code(FAIL_SERVER_CONNECTION_ERROR);
            ret.__set_msg(e.what());
            (this->thrift_client_).reset();
        } catch (...) {
            state_ = CONNECTION_UNHEALTHY;
            ret.__set_code(FAIL_SERVER_CONNECTION_ERROR);
            ret.__set_msg("Unkown exception.");
            (this->thrift_client_).reset();
        }
        if (ret.key.empty()) {
            ret.__set_key(msg.key);
        }
        return ret;
    }

     DelayResult CarreraConnection::Send(const DelayMessage &msg, int64_t proxy_timeout) {
        DelayResult ret;
        try {
            if (NULL == this->thrift_client_) {
                this->thrift_client_ = CreateThriftClient();
                if (NULL == this->thrift_client_) {
                    ret.__set_code(FAIL_SERVER_CONNECTION_ERROR);
                    ret.__set_msg("Illegal ipport or open transport failed.");
                    return ret;
                }
            }
            this->thrift_client_->sendDelaySync(ret, msg, proxy_timeout);
        } catch (TException &e) {
            state_ = CONNECTION_UNHEALTHY;
            ret.__set_code(FAIL_SERVER_CONNECTION_ERROR);
            ret.__set_msg(e.what());
            (this->thrift_client_).reset();
        } catch (std::exception &e) {
            state_ = CONNECTION_UNHEALTHY;
            ret.__set_code(FAIL_SERVER_CONNECTION_ERROR);
            ret.__set_msg(e.what());
            (this->thrift_client_).reset();
        } catch (...) {
            state_ = CONNECTION_UNHEALTHY;
            ret.__set_code(FAIL_SERVER_CONNECTION_ERROR);
            ret.__set_msg("Unkown exception.");
            (this->thrift_client_).reset();
        }
        return ret;
    }

}

